File: bfs/min_buses_route/min_buses_route.py

def find_min_buses_route(routes: list[list[int]], source: int, target: int) -> int:  # LeetCode Q.815.
    if source == target:  # Base case.
        return 0

    source_routes, target_routes = set(), set()
    routes_intersections: dict[int, set[int]] = dict()

    total_routes = len(routes)
    for route_1, route_1_stops in enumerate(routes):
        if source in route_1_stops:
            source_routes.add(route_1)

        if target in route_1_stops:
            target_routes.add(route_1)

        if route_1 not in routes_intersections.keys():
            routes_intersections.update({route_1: set()})

        for route_2 in range(route_1 + 1, total_routes):
            route_2_stops = routes[route_2]
            if set(route_1_stops) & set(route_2_stops) != set():
                routes_intersections[route_1].add(route_2)
                if route_2 not in routes_intersections.keys():
                    routes_intersections.update({route_2: set()})
                routes_intersections[route_2].add(route_1)

    if target_routes:  # Target stop has bus routes.
        taken_routes = set()
        # Format: (current route, buses taken).
        queue = [(route, 1) for route in source_routes]

        while queue:
            current_route, buses_taken = queue.pop(0)
            if target in routes[current_route]:
                return buses_taken

            if current_route not in taken_routes:
                taken_routes.add(current_route)
                buses_taken += 1
                if current_route in routes_intersections.keys():
                    for intersection_route in routes_intersections[current_route]:
                        if intersection_route not in taken_routes:
                            queue.append((intersection_route, buses_taken))

    return -1

File: bfs/min_buses_route/min_buses_route.cpp
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <queue>
using namespace std;

int numBusesToDestination(vector<vector<int>> &routes, int source, int target)
{
    if (source == target) // Base case: source is target.
        return 0;

    unordered_set<int> source_routes, target_routes;

    // Keys: stations. Values: crossed routes.
    unordered_map<int, vector<int>> stations2routes;
    for (int idx = 0; idx < routes.size(); idx++)
    {
        for (auto station : routes[idx])
        {
            if (station == source)
                source_routes.insert(idx + 1); // Route num = idx + 1.

            if (station == target)
                target_routes.insert(idx + 1); // Route num = idx + 1.

            if (stations2routes.find(station) == stations2routes.end())
                stations2routes[station] = {};
            stations2routes[station].push_back(idx + 1); // Route num = idx + 1.
        }
    }

    if (source_routes.empty() || target_routes.empty())
        return -1; // Base case: source or target lacks routes.

    unordered_set<int> visited_routes, visited_stations;
    queue<pair<int, int>> queue; // Format: {route num, count of taken routes}.
    for (auto route : source_routes)
        queue.push({route, 1});

    while (!queue.empty())
    {
        auto [route, buses_taken] = queue.front();
        visited_routes.insert(route);
        queue.pop();

        if (target_routes.find(route) != target_routes.end())
            return buses_taken;

        buses_taken += 1;
        for (auto station : routes[route - 1]) // Route num = idx + 1.
        {
            if (visited_stations.find(station) == visited_stations.end())
            {
                visited_stations.insert(station);

                for (auto crossed_route : stations2routes[station])
                {
                    if (visited_routes.find(crossed_route) == visited_routes.end())
                        queue.push({crossed_route, buses_taken});
                }
            }
        }
    }

    return -1;
}
